packages/php-wasm/universal/src/lib/php-request-handler.ts

```diff
@@ -598,6 +662,232 @@ export class PHPRequestHandler implements AsyncDisposable {
+			 *
+			 * Consider this Apache example from above:
+			 *
+			 * – Document Root: /var/www/html
+			 * – Script file:   /var/www/html/subdir/script.php
+			 * – Rewrite rule:  ^api/v1/user/([0-9]+)$ /subdir/script.php?endpoint=user&id=$1 [L,QSA]
+			 * – Requesting     /api/v1/user/123
+			 *
+			 *   $_SERVER['PHP_SELF'] = "/subdir/script.php"
+			 *
+			 * So, on the face value, this is a filesystem path. However, see
+			 * what happens if we slightly modify that rewrite rule to:
+			 *
+			 * – Rewrite rule:  ^api/v1/user/([0-9]+)$ /subdir/script.php/next.php
+			 *                                                           ^^^^^^^^^
+			 * – Requesting     /api/v1/user/123
+			 *
+			 *   $_SERVER['PHP_SELF'] = "/subdir/script.php/next.php"
+			 *
+			 * So:
+			 * * PHP_SELF is not sourced from the filesystem path.
+			 * * PHP_SELF is sourced from the final request URL after the
+			 *   rewrite rules have been applied.
+			 */
+			$_SERVER['PHP_SELF'] = rewrittenRequestUrl.pathname;
+
+			/**
+			 * PATH_INFO
+			 *
+			 * > Contains any client-provided pathname information trailing the actual
+			 * > script filename but preceding the query string, if available. For instance,
+			 * > if the current script was accessed via the URI http://www.example.com/php/path_info.php/some/stuff?foo=bar,
+			 * > then $_SERVER['PATH_INFO'] would contain /some/stuff.
+			 *
+			 * This **does not** include the query string.
+			 *
+			 * @see https://www.php.net/manual/en/reserved.variables.server.php#:~:text=PATH_INFO
+			 */
+			if ($_SERVER['REQUEST_URI'].startsWith($_SERVER['SCRIPT_NAME'])) {
+				$_SERVER['PATH_INFO'] = $_SERVER['REQUEST_URI'].substring(
+					$_SERVER['SCRIPT_NAME'].length
+				);
+				// Remove the query string if present.
+				if ($_SERVER['PATH_INFO'].includes('?')) {
+					$_SERVER['PATH_INFO'] = $_SERVER['PATH_INFO'].substring(
+						0,
+						$_SERVER['PATH_INFO'].indexOf('?')
+					);
+				}
+			}
+		}
+
+		/**
+		 * QUERY_STRING
+		 *
+		 * The query string from the original and rewritten request URLs.
+		 * Does not include the leading question mark.
+		 *
+		 * Note it contains all the query parameters from the original
+		 * URL merged with the new parameters from the rewritten request URLs.
+		 *
+		 * Example:
+		 *    – Original request URL: /pretty/url?foo=bar&page=different-value
+		 *    – Rewritten request URL: /pretty/url?page=pretty
+		 *    – QUERY_STRING: page=pretty&foo=bar&page=different-value
+		 */
+		$_SERVER['QUERY_STRING'] = rewrittenRequestUrl.search.substring(1);
+
+		/**
+		 * There's a few relevant entries we are NOT setting here:
+		 *
+		 *    – SCRIPT_FILENAME: Absolute path to the script file. It is set by
+		 *      php_wasm.c.
+		 *    – REDIRECT_STATUS: Apache sets it, but it's optional so we skip it.
+		 *    – REDIRECT_URL: Apache sets it, but it's optional so we skip it.
+		 *    – REDIRECT_QUERY_STRING: Apache sets it, but it's optional so we skip it.
+		 */
```

Comment:
I really appreciate the clarity and thoroughness in this entire large block.

Reply:
Thank you! I've got confused by this `$_SERVER` business so many times, I want to make sure the next person has the full information needed to make changes readily available right in the code.